task_processing/plugins/kubernetes/kubernetes_pod_executor.py

```diff
@@ -173,25 +191,44 @@ def kill(self, task_id: str) -> bool:
         # TODO(TASKPROC-242): actually handle termination events
         logger.info(f"Attempting to terminate Pod: {task_id}")
-        try:
-            status: V1Status = self.kube_client.core.delete_namespaced_pod(
-                name=task_id,
-                namespace=self.namespace,
-                # attempt to delete immediately - Pods launched by task_processing
-                # shouldn't need time to clean-up/drain
-                grace_period_seconds=0,
-                # this is the default, but explcitly request background deletion of releated objects
-                # see: https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/
-                propagation_policy="Background"
-            )
-        except Exception:
-            logger.exception(f"Failed to request termination for Pod: {task_id}")
-            return False
-
-        # this is not ideal, but the k8s clientlib returns the status of the request as a string
-        # that is either "Success" or "Failure" - we could potentially use `code` instead
-        # but it's not exactly documented what HTTP return codes will be used
-        return status.status == "Success"
+        # only retry once, reloading multiple times isn't likely to do much
+        retries = 1
+        while retries >= 0:
+            try:
+                status: V1Status = self.kube_client.core.delete_namespaced_pod(
+                    name=task_id,
+                    namespace=self.namespace,
+                    # attempt to delete immediately - Pods launched by task_processing
+                    # shouldn't need time to clean-up/drain
+                    grace_period_seconds=0,
+                    # this is the default, but explcitly request background deletion of releated
+                    # objects. see:
+                    # https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/
+                    propagation_policy="Background"
+                )
+                # this is not ideal, but the k8s clientlib returns the status of the request as a
+                # string that is either "Success" or "Failure" - we could potentially use `code`
+                # instead but it's not exactly documented what HTTP return codes will be used
+                return status.status == "Success"
+            except ApiException as e:
+                if e.status == HTTPStatus.UNAUTHORIZED.value:
+                    logger.info(
+                        "Recieved UNAUTHORIZED response from apiserver - assuming certs have "
+                        "expired and reloading."
+                    )
+                    self.kube_client.reload_kubeconfig()
+                else:
+                    logger.exception(
+                        f"Failed to request termination for Pod {task_id} due to unhandled API "
+                        "exception, retrying."
+                    )
```

Comment:
a matter of taste, but this being repeated few times, looks like it could be factored out into a function

Reply:
I moved things over to the KubeClient class so that there's less logic in the plugin code itself, but I wasn't able to get rid of 100% of the duplication - but at least now it's just the general structure that's duplicated and not the exact same code